merlin/core/dispatch.py

```diff
@@ -217,50 +213,78 @@ def pd_convert_hex(x):
 
 def random_state(seed, like_df=None):
     """Dispatch for numpy.random.RandomState"""
-    if not HAS_GPU or isinstance(like_df, (pd.DataFrame, pd.Series)):
+    if isinstance(like_df, (pd.DataFrame, pd.Series)):
         return np.random.RandomState(seed)
-    else:
+    elif cudf and isinstance(like_df, (cudf.DataFrame, cudf.Series)):
         return cp.random.RandomState(seed)
+    else:
+        return ValueError(
+            "Unsupported dataframe type: "
+            f"{type(like_df)}"
+            " Expected either a pandas or cudf DataFrame or Series."
+        )
 
 
 def arange(size, like_df=None, dtype=None):
     """Dispatch for numpy.arange"""
-    if not HAS_GPU or isinstance(like_df, (np.ndarray, pd.DataFrame, pd.Series)):
+    if isinstance(like_df, (np.ndarray, pd.DataFrame, pd.Series)):
         return np.arange(size, dtype=dtype)
-    else:
+    elif cudf and isinstance(like_df, (cp.ndarray, cudf.DataFrame, cudf.Series)):
```

Comment:
The code in dispatch.py was written in this (admittedly somewhat weird) fashion to avoid directly referencing GPU-only libraries that won’t be available on CPU. I think the path forward here is likely to apply ‘@lazy_single dispatch’ so that we register a GPU version of the function only when CuPy/cuDF are available

Reply:
It turned out to be easier to keep this PR representing the addition of a test for GPU environment with no visible devices (and corresponding fixes to the tensor table tests). Moved the dispatch changes to #244 .

Reply:
I also added another piece of information in the description of that PR, which is that prior to #211 , the `HAS_GPU` variable represented both the detection of a GPU device and the successful import of cudf. Assuming a successful import of cudf tells us both whether cudf is installed and whether or not there is at least one GPU device then the checking of the cudf  before every usage is roughly equivalent to what we were doing prior to #211

Reply:
Ah yeah, I hadn't considered that checking `cudf` also implicitly checks `cupy` 👍🏻

Reply:
> the isinstance check uses both cudf and cp but only checks if cudf exists first

If cudf depends on cupy, then this might be ok, since the presence of cudf implies cupy is also available?

Reply:
Agreed ✅